### PR TITLE
logcheck: check usage of format specifier in structured log functions

### LINKED ISSUE #218: logcheck: Add check for format string usage
/kind feature

**Describe the solution you'd like**
Improve `logcheck` tool to check for format string usage in structured logging functions like `InfoS`, `ErrorS`.

**Anything else you would like to add:**
We are reporting error if found usage of unstructured klog functions [here](https://github.com/kubernetes/klog/blob/master/hack/tools/logcheck/main.go#L86). In similar way we need to report usage of format strings, the tool should report error on following line.
`klog.InfoS("pod %s started", podName)`

related to: https://github.com/kubernetes/kubernetes/issues/98975

### PR DESCRIPTION
**Pull request by umangachapagain**
Signed-off-by: Umanga Chapagain <chapagainumanga@gmail.com>

**What this PR does / why we need it**:
logcheck: check usage of format specifier in structured logging functions.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #218 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: hack/tools/logcheck/main.go [hunk 1]
```diff
@@ -81,13 +82,18 @@ func checkForFunctionExpr(fun ast.Expr, pass *analysis.Pass) {
...
-			})
+		if ok && pName.Name == "klog" {
+			isUnstructured := isUnstructured((fName))
+			if isUnstructured {
+				msg := fmt.Sprintf("unstructured logging function %q should not be used", fName)
+				pass.Report(analysis.Diagnostic{
+					Pos:     fun.Pos(),
+					Message: msg,
+				})
+			}
```
Comment (adisky) [inline hack/tools/logcheck/main.go:93]: why not `else` here?

File: hack/tools/logcheck/main.go [hunk 2]
```diff
@@ -110,3 +116,35 @@ func isUnstructured(fName string) bool {
 
 	return false
 }
+
+func checkForFormatSpecifier(expr *ast.CallExpr, pass *analysis.Pass) {
+	if selExpr, ok := expr.Fun.(*ast.SelectorExpr); ok {
+		// extracting function Name like Infof
+		fName := selExpr.Sel.Name
+		if hasFormatSpecifier(expr.Args) {
+			msg := fmt.Sprintf("structured logging function %q should not use format specifier", fName)
```
Comment (adisky) [inline hack/tools/logcheck/main.go:125]: can we print format specifier here?

**Comment by adisky** [comment]
Thanks @umangachapagain for working on this. It is catching for format specifiers \o/!!
some suggestions inline

File: hack/tools/logcheck/main.go (hunk 2 above)
Comment (umangachapagain) [reply hack/tools/logcheck/main.go]: Done.

File: hack/tools/logcheck/main.go (hunk 1 above)
Comment (umangachapagain) [reply hack/tools/logcheck/main.go]: I don't know why I did that 🤦🏽

**Comment by umangachapagain** [comment]
Usage like `%.2f` aren't caught. Should we have an issue for that?
It'd need more than just string matching IMO.

**Comment by serathius** [comment]
I think it's ok to skip handling all possible format specifiers. Main purpose of this issue is to catch cases where someone incorrectly migrates log line without removing formatting. 
Overall looks good to me, could you run it against `pkg/kubelet` and `cmd/kubelet` directories in https://github.com/kubernetes/kubernetes and see if there are any problems ?

**Comment by umangachapagain** [comment]
> Overall looks good to me, could you run it against `pkg/kubelet` and `cmd/kubelet` directories in https://github.com/kubernetes/kubernetes and see if there are any problems ?
>
Ran test with latest build.
```
$ logcheck ./pkg/kubelet/... ./cmd/kubelet/...                                                                                                                                                                                            
/go/src/k8s.io/kubernetes/pkg/kubelet/network/dns/dns.go:126:3: Additional arguments to ErrorS should always be Key Value pairs. Please check if there is any key or value missing.
/go/src/k8s.io/kubernetes/pkg/kubelet/network/dns/dns.go:136:3: Additional arguments to ErrorS should always be Key Value pairs. Please check if there is any key or value missing.
```
So no issues with format specifiers at least.

File: hack/tools/logcheck/testdata/data.go [hunk 3]
```diff
@@ -23,22 +23,25 @@ import (
...
+	klog.WarningDepth(1, "test log")    // want `unstructured logging function "WarningDepth" should not be used`
+	klog.Error("test log")              // want `unstructured logging function "Error" should not be used`
+	klog.Errorf("test log")             // want `unstructured logging function "Errorf" should not be used`
+	klog.Errorln("test log")            // want `unstructured logging function "Errorln" should not be used`
+	klog.ErrorDepth(1, "test log")      // want `unstructured logging function "ErrorDepth" should not be used`
+	klog.Fatal("test log")              // want `unstructured logging function "Fatal" should not be used`
+	klog.Fatalf("test log")             // want `unstructured logging function "Fatalf" should not be used`
+	klog.Fatalln("test log")            // want `unstructured logging function "Fatalln" should not be used`
+	klog.FatalDepth(1, "test log")      // want `unstructured logging function "FatalDepth" should not be used`
+	klog.InfoS("test: %s", "testname")  // want `structured logging function "InfoS" should not use format specifier "%s"`
```
Comment (adisky) [inline hack/tools/logcheck/testdata/data.go:42]: Move this calls to `printStructuredLog()` function

Comment (umangachapagain) [reply hack/tools/logcheck/testdata/data.go]: Done.

**Comment by adisky** [comment]
LGTM!! Thanks @umangachapagain 
/cc @serathius

**Pull request merged**